File: Braintree/src/main/java/com/braintreepayments/api/exceptions/ErrorWithResponse.java
package com.braintreepayments.api.exceptions;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.Nullable;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

/**
 *  Error container returned when the Braintree server returns a 422 Unprocessible Entity.
 *  A 422 occurs when a request is properly formed, but the server was unable to take the requested
 *  action due to bad user data.
 *
 *  ErrorWithResponse parses the server's error response and exposes the errors.
 */
public class ErrorWithResponse extends Exception implements Parcelable {

    private static final String ERROR_KEY = "error";
    private static final String MESSAGE_KEY = "message";
    private static final String FIELD_ERRORS_KEY = "fieldErrors";

    private int mStatusCode;
    private String mMessage;
    private String mOriginalResponse;
    private List<BraintreeError> mFieldErrors;

    public ErrorWithResponse(int statusCode, String jsonString) {
        mStatusCode = statusCode;
        mOriginalResponse = jsonString;

        try {
            parseJson(jsonString);
        } catch (JSONException e) {
            mMessage = "Parsing error response failed";
            mFieldErrors = new ArrayList<>();
        }
    }

    private ErrorWithResponse() {}

    public static ErrorWithResponse fromJson(String json) throws JSONException {
        ErrorWithResponse errorWithResponse = new ErrorWithResponse();
        errorWithResponse.mOriginalResponse = json;
        errorWithResponse.parseJson(json);

        return errorWithResponse;
    }

    private void parseJson(String jsonString) throws JSONException {
        JSONObject json = new JSONObject(jsonString);
        mMessage = json.getJSONObject(ERROR_KEY).getString(MESSAGE_KEY);
        mFieldErrors = BraintreeError.fromJsonArray(json.optJSONArray(FIELD_ERRORS_KEY));
    }

    /**
     * @return HTTP status code from the Braintree gateway.
     */
    public int getStatusCode() {
        return mStatusCode;
    }

    /**
     * @return Human readable top level summary of the error.
     */
    @Override
    public String getMessage() {
        return mMessage;
    }

    /**
     * @return The full error response as a {@link String}.
     */
    public String getErrorResponse() {
        return mOriginalResponse;
    }

    /**
     * @return All the field errors.
     */
    public List<BraintreeError> getFieldErrors() {
        return mFieldErrors;
    }

    /**
     * Method to extract an error for an individual field, e.g. creditCard, customer, etc.
     *
     * @param field Name of the field desired, expected to be in camelCase.
     * @return {@link BraintreeError} for the field searched, or {@code null} if not found.
    */
    @Nullable
    public BraintreeError errorFor(String field) {
        BraintreeError returnError;
        if(mFieldErrors != null) {
            for (BraintreeError error : mFieldErrors) {
                if (error.getField().equals(field)) {
                    return error;
                } else if (error.getFieldErrors() != null) {
                    returnError = error.errorFor(field);
                    if (returnError != null) {
                        return returnError;
                    }
                }
            }
        }
        return null;
    }

    @Override
    public String toString() {
        return "ErrorWithResponse (" + mStatusCode + "): " + mMessage + "\n" +
                mFieldErrors.toString();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(mStatusCode);
        dest.writeString(mMessage);
        dest.writeString(mOriginalResponse);
        dest.writeTypedList(mFieldErrors);
    }

    protected ErrorWithResponse(Parcel in) {
        mStatusCode = in.readInt();
        mMessage = in.readString();
        mOriginalResponse = in.readString();
        mFieldErrors = in.createTypedArrayList(BraintreeError.CREATOR);
    }

    public static final Creator<ErrorWithResponse> CREATOR = new Creator<ErrorWithResponse>() {
        public ErrorWithResponse createFromParcel(Parcel source) {
            return new ErrorWithResponse(source);
        }

        public ErrorWithResponse[] newArray(int size) {
            return new ErrorWithResponse[size];
        }
    };
}

File: BraintreeCore/src/main/java/com/braintreepayments/api/core/ErrorWithResponse.kt
package com.braintreepayments.api.core

import android.os.Parcelable
import android.util.Log
import androidx.annotation.RestrictTo
import com.braintreepayments.api.core.GraphQLConstants.ErrorMessages
import com.braintreepayments.api.sharedutils.LoggingUtils
import kotlinx.parcelize.Parcelize
import org.json.JSONException
import org.json.JSONObject

/**
 * Error container returned when the Braintree server returns a 422 Unprocessible Entity.
 * A 422 occurs when a request is properly formed, but the server was unable to take the requested
 * action due to bad user data.
 *
 * ErrorWithResponse parses the server's error response and exposes the errors.
 *
 * @property statusCode HTTP status code from the Braintree gateway.
 * @property message Human readable top level summary of the error.
 * @property errorResponse The full error response as a [String].
 * @property fieldErrors All the field errors.
 */
@Parcelize
data class ErrorWithResponse internal constructor(
    var statusCode: Int = 0,
    var errorResponse: String?,
    var fieldErrors: List<BraintreeError>? = null,
    override var message: String? = null,
) : Exception(), Parcelable {

    init {
        try {
            parseJson(errorResponse)
        } catch (e: JSONException) {
            Log.d(LoggingUtils.TAG, e.message.toString())
            fieldErrors = ArrayList()
        }
    }

    @Throws(JSONException::class)
    private fun parseJson(jsonString: String?) {
        jsonString?.let { JSONObject(it) }?.let { json ->
            message = json.optJSONObject(ERROR_KEY)
                ?.let { jsonObject ->
                    jsonObject.optString(MESSAGE_KEY, "").takeIf { it.isNotEmpty() }
                        ?: jsonObject.optString(DEV_MESSAGE_KEY, "")
                } ?: throw JSONException("Error key not found in JSON")
            fieldErrors = BraintreeError.fromJsonArray(json.optJSONArray(FIELD_ERRORS_KEY))
        }
    }

    /**
     * Method to extract an error for an individual field, e.g. creditCard, customer, etc.
     *
     * @param field Name of the field desired, expected to be in camelCase.
     * @return [BraintreeError] for the field searched, or `null` if not found.
     */
    fun errorFor(field: String): BraintreeError? {
        if (fieldErrors == null) return null

        for (error in fieldErrors!!) {
            if (error.field == field) {
                return error
            } else if (error.fieldErrors != null) {
                val returnError = error.errorFor(field)
                if (returnError != null) {
                    return returnError
                }
            }
        }
        return null
    }

    override fun toString(): String {
        return """
            ErrorWithResponse ($statusCode): $message
            $fieldErrors
        """.trimIndent()
    }

    companion object {
        private const val ERROR_KEY = "error"
        private const val MESSAGE_KEY = "message"
        private const val DEV_MESSAGE_KEY = "developer_message"
        private const val FIELD_ERRORS_KEY = "fieldErrors"
        private const val GRAPHQL_ERROR_CODE = 422

        /**
         * @suppress
         */
        @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
        @Throws(JSONException::class)
        @JvmStatic
        fun fromJson(json: String?) = ErrorWithResponse(errorResponse = json).apply {
            parseJson(json)
        }

        internal fun fromGraphQLJson(json: String?): ErrorWithResponse {
            val errorWithResponse = ErrorWithResponse(
                errorResponse = json,
                statusCode = GRAPHQL_ERROR_CODE
            )

            try {
                val errors = json
                    ?.let { JSONObject(it) }
                    ?.optJSONArray(GraphQLConstants.Keys.ERRORS)
                    ?: throw JSONException("Errors key not found in JSON")
                errorWithResponse.fieldErrors = BraintreeError.fromGraphQLJsonArray(errors)

                val fieldErrorsEmpty = errorWithResponse.fieldErrors?.isEmpty() ?: true
                errorWithResponse.message = if (fieldErrorsEmpty) {
                    errors.getJSONObject(0)?.getString(GraphQLConstants.Keys.MESSAGE)
                } else {
                    ErrorMessages.USER
                }
            } catch (e: JSONException) {
                errorWithResponse.apply {
                    Log.d(LoggingUtils.TAG, e.message.toString())
                    fieldErrors = ArrayList()
                }
            }
            return errorWithResponse
        }
    }
}
